### ISSUE TITLE
web UI: use gopherjs (?)

### ISSUE THREAD

**Issue by mpl** [camlistore/camlistore#798]
There is a will (primarily from @bradfitz and I) to start using gopherjs (gopherjs.org) for future work on the web UI.

Some of the problems related to that move have been discussed in issue #607 , but I'm creating this new issue for further discussion since the original intent of 607 was not the same (even though the issues are related imo).

A recap of the existing gopherjs CLs and experiments:
https://camlistore-review.googlesource.com/#/c/5816/
https://camlistore-review.googlesource.com/#/c/6056/
https://camlistore-review.googlesource.com/#/c/6136/
https://camlistore-review.googlesource.com/#/c/6608/

**Comment by mpl** [comment]
Noteworthy, and somewhat related to the React problem: I've just discovered this project: https://github.com/gopherjs/vecty which seems to be getting contributions from the gopherjs maintainer.

**Comment by DanielHeath** [comment]
When comparing filesizes, it's important to bear in mind that there's a lot more we could do to optimize the camlistore frontend. Unfortunately most of the tools for that stuff are not easily ported to go - e.g. https://developers.google.com/closure/compiler/ .

**Comment by mpl** [comment]
@DanielHeath it doesn't have to be in Go to be called from Go. we actually used to run the closure compiler before we started using React.
And I've retried it recently too:
https://github.com/camlistore/camlistore/issues/607#issuecomment-220742124

**Comment by DanielHeath** [comment]
I've tried to do frontend stuff in gopherjs recently; if anyone is thinking of embarking on this path it's worth noting that you will have to write a lot of wrapper boilerplate. 

Libraries exist to do it but generally only support the subset of properties the author happened to need.

**Comment by mpl** [comment]
@DanielHeath from what I tested in e.g. https://camlistore-review.googlesource.com/#/c/6608/ it didn't feel so bad, and even if you're right I'm not worried about having to write more Go code tbh.

On the other hand, the output size may turn out to be enough of a problem to be a show stopper for us. :-(

With something like a ~4MB gopherjs.js, it takes ~10s to load the index page of the web UI on my remote server. As it is supposed to be a long lived app, I personally don't mind it so much, but it's something we have to weight seriously.

**Comment by cliveb** [comment]
Tree shaking like in Dart to make gopherjs output size smaller discussed
and closed. With comment from @neelance <https://github.com/neelance>
https://github.com/gopherjs/gopherjs/issues/259

**Comment by edrex** [comment]
I'm planning to explore TS/ESM/third party apps in parallel [here](https://github.com/edrex/camlapp).

**Comment by cliveb** [comment]
Perhaps some insight on tree-shaking from Ian on the go team:

It's not the compiler. It's the build system. One of the design
goals for Go was to support very large programs, such as the programs
written at Google.

We did that to help programming in the large, recognizing
that it would cause some pain for programming in the small. Every
language is full of tradeoffs.

http://grokbase.com/p/gg/golang-nuts/149ycfxqeb/go-nuts-re-go-complains-too-much

**Comment by mpl** [comment]
I've been working on an alternative avenue. Since the publisher is not as critical as the web UI, and is always in need of some polishing, I've added infinite scrolling to the publisher with gopherjs.

I need to test it a bit more, but it looks like it's working nicely enough so far. And again, the experiment seems to confirm that the js code size does not matter that much for a long lived page which has to download much data anyway.

So once I've squashed all the CLs together, I think I could propose one that uses gopherjs only for the publisher so far. At least all the setup will be ready if/when we want to enable it for the web UI too.

https://camlistore-review.googlesource.com/6796

**Comment by cliveb** [comment]
On Full Stack Development and Go an upcoming talk has taken an infinite
scrolling approach for Stardew Farms publishing. Apparently with node.js
rather than gopher.js. I will try to learn why. http://stardew.farm/

http://www.meetup.com/golang/events/231455969/

**Comment by mpl** [comment]
Infinite scrolling in publisher: https://camlistore-review.googlesource.com/6886

**Comment by mpl** [comment]
@edrex something you might be interested in: https://github.com/gopherjs/gopherjs/issues/524

**Comment by paralin** [comment]
@edrex Yes, TypeScript is awesome, use it. My interest in using Gopher right now is for a web game I'm working on with a small team where we want to have symmetric game logic between the browser and the server. We need performance on the server, so any kind of JavaScript is sucky. Thus Go.

Use Webpack, Angular2 if you can. I think this is the best option right now.

See https://github.com/AngularClass/angular2-webpack-starter

**Comment by mpl** [comment]
Paul Jolly is currently working on some react bindings for gopherjs, so we should keep an eye on that as well.
https://github.com/myitcv/gopherjs/wiki

**Comment by shurcooL** [comment]
I'm not seeing it mentioned here, so I will. There are also some React bindings by @bep at https://github.com/bep/gr.

**Comment by mpl** [comment]
https://camlistore-review.googlesource.com/9086

**Comment by mpl** [comment]
@myitcv @edrex ping ^^
